core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala

```diff
@@ -1,22 +1,90 @@
+import org.broadinstitute.dsde.rawls.entities.exceptions.DataEntityException
+import org.broadinstitute.dsde.rawls.model.DataReferenceModelJsonSupport.TerraDataRepoSnapshotRequestFormat
+import org.broadinstitute.dsde.rawls.model.{DataReferenceName, TerraDataRepoSnapshotRequest}
+import spray.json._
 
 import scala.concurrent.ExecutionContext
 import scala.reflect.runtime.universe._
+import scala.util.{Failure, Success, Try}
 
 class DataRepoEntityProviderBuilder(workspaceManagerDAO: WorkspaceManagerDAO, dataRepoDAO: DataRepoDAO,
                                        samDAO: SamDAO, bqServiceFactory: GoogleBigQueryServiceFactory)
                                    (implicit protected val executionContext: ExecutionContext)
-  extends EntityProviderBuilder[DataRepoEntityProvider] {
+  extends EntityProviderBuilder[DataRepoEntityProvider] with LazyLogging {
 
   override def builds: TypeTag[DataRepoEntityProvider] = typeTag[DataRepoEntityProvider]
 
-  override def build(requestArguments: EntityRequestArguments): DataRepoEntityProvider = {
-    new DataRepoEntityProvider(requestArguments, workspaceManagerDAO, dataRepoDAO, samDAO, bqServiceFactory)
+  override def build(requestArguments: EntityRequestArguments): Try[DataRepoEntityProvider] = {
+    for {
+      dataReferenceName <- requestArguments.dataReference.toRight(new DataEntityException("data reference must be defined for this provider")).toTry
+
+      // get snapshot UUID from data reference name
+      snapshotId <- Try(lookupSnapshotForName(dataReferenceName, requestArguments))
+
+      // contact TDR to describe the snapshot
+      snapshotModel <- Try(dataRepoDAO.getSnapshot(snapshotId, requestArguments.userInfo.accessToken)).recoverWith {
+        case notFound: DatarepoApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
+          Failure(new DataEntityException(s"Snapshot id $snapshotId does not exist or you do not have access", notFound))
+        case forbidden: DatarepoApiException if forbidden.getCode == StatusCodes.Forbidden.intValue =>
+          Failure(new DataEntityException(s"Snapshot id $snapshotId exists but access was denied", forbidden))
+      }
+    } yield new DataRepoEntityProvider(snapshotModel, requestArguments, samDAO, bqServiceFactory)
+  }
+
+  private[datarepo] def lookupSnapshotForName(dataReferenceName: DataReferenceName, requestArguments: EntityRequestArguments): UUID = {
+    // contact WSM to retrieve the data reference specified in the request
+    val dataRefTry = Try(workspaceManagerDAO.getDataReferenceByName(UUID.fromString(requestArguments.workspace.workspaceId),
+      ReferenceTypeEnum.DATAREPOSNAPSHOT.getValue,
+      dataReferenceName,
+      requestArguments.userInfo.accessToken)).recoverWith {
+
+      case notFound: WorkspaceApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
+        Failure(new DataEntityException(s"Reference name ${dataReferenceName.value} does not exist.", notFound))
```

Comment:
Would it add more clarity here to say the Reference does not exist in the Workspace?